File: src/test_programs/C/test_geomgrad_xyz.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "openmmpol.h"

int countLines(char *fin){
    FILE *fp = fopen(fin, "r");
    
    char c;
    int lines = 1;

    if(fp == NULL) return 0;

    do{
        c = fgetc(fp);
        if(c == '\n') lines++;
    }while(c != EOF);

    fclose(fp);
  
    return lines - 1;
}

double **read_ef(char *fin){
    double **ef;
    
    int pol_atoms = countLines(fin);

    ef = (double **) malloc(sizeof(double *) * 3 * pol_atoms);

    FILE *fp = fopen(fin, "r");
    for(int i =0; i < pol_atoms; i++){
        ef[i] = (double *) malloc(sizeof(double) * 3);
        fscanf(fp, "%lf %lf %lf", &(ef[i][0]), &(ef[i][1]),  &(ef[i][2]));
    }
    fclose(fp);

    return ef;
}

int ana_grd_print(OMMP_SYSTEM_PRT sys,
                  void (*grad_f)(OMMP_SYSTEM_PRT, double *), 
                  FILE *fp, const char *name){
    
    int mm_atoms = ommp_get_mm_atoms(sys);
    
    double *_grad_ana = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    double **grad_ana = (double **) malloc(sizeof(double *) * mm_atoms);

    for(int i = 0; i < mm_atoms; i++){
        grad_ana[i] = &(_grad_ana[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana[i][j] = 0.0;
    }

    grad_f(sys, _grad_ana);
   
    fprintf(fp, "Grad %s\n", name);
    
    for(int i = 0; i < mm_atoms; i++){
        for(int j=0; j < 3; j++){
            fprintf(fp, "%+12.8g ", grad_ana[i][j]*AU2KCALMOL*ANG2AU);
        }
        fprintf(fp, "\n");
    }
    fprintf(fp, "\n");
}

int main(int argc, char **argv){
    if(argc != 4){
        printf("Syntax expected\n");
        printf("    $ test_geomgrad_xyz.exe <XYZ FILE> <PRM FILE> <OUTPUT FILE>\n");
        return 0;
    }
    
    int pol_atoms, mm_atoms, rc=0;
    double **grad_num, *_grad_num, **grad_ana, *_grad_ana;
    double delta, Mdelta;

    OMMP_SYSTEM_PRT my_system = ommp_init_xyz(argv[1], argv[2]);
    ommp_set_verbose(OMMP_VERBOSE_NONE);
    pol_atoms = ommp_get_pol_atoms(my_system);
    mm_atoms = ommp_get_mm_atoms(my_system);
    
    //electric_field = (double *) malloc(sizeof(double) * 3 * pol_atoms);
    //polar_mm = (int32_t *) ommp_get_polar_mm(my_system);

    _grad_num = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    grad_num = (double **) malloc(sizeof(double *) * mm_atoms);
    _grad_ana = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    grad_ana = (double **) malloc(sizeof(double *) * mm_atoms);
    for(int i = 0; i < mm_atoms; i++){
        grad_ana[i] = &(_grad_ana[i*3]);
        grad_num[i] = &(_grad_num[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana[i][j] = grad_num[i][j] = 0.0;
    }
   
    FILE *fp = fopen(argv[3], "w+");
   
    //ana_grd_print(my_system, ommp_full_geomgrad, fp, "ETOT");
    
    ana_grd_print(my_system, ommp_vdw_geomgrad, fp, "EV");
    ana_grd_print(my_system, ommp_fixedelec_geomgrad, fp, "EM");
    ana_grd_print(my_system, ommp_polelec_geomgrad, fp, "EP");
    ana_grd_print(my_system, ommp_bond_geomgrad, fp, "EB");
    ana_grd_print(my_system, ommp_angle_geomgrad, fp, "EA");
    ana_grd_print(my_system, ommp_urey_geomgrad, fp, "EUB");
    ana_grd_print(my_system, ommp_torsion_geomgrad, fp, "ET");
    ana_grd_print(my_system, ommp_imptorsion_geomgrad, fp, "EIT");
    ana_grd_print(my_system, ommp_strbnd_geomgrad, fp, "EBA");
    ana_grd_print(my_system, ommp_angtor_geomgrad, fp, "EAT");
    ana_grd_print(my_system, ommp_opb_geomgrad, fp, "EOPB");
    ana_grd_print(my_system, ommp_strtor_geomgrad, fp, "EBT");
    ana_grd_print(my_system, ommp_tortor_geomgrad, fp, "ETT");
    ana_grd_print(my_system, ommp_pitors_geomgrad, fp, "EPT");
    
    fclose(fp);
    ommp_terminate(my_system);
     
    return rc;
}

File: src/test_programs/F03/test_geomgrad_xyz.f90
module anagg_print
    abstract interface
    subroutine grad_term(s, grad)
        use mod_mmpol, only: ommp_system
        use mod_memory, only: rp
        type(ommp_system), intent(inout), target :: s
        real(rp), intent(out) :: grad(3,s%top%mm_atoms)
    end subroutine
    end interface

    contains

    subroutine ana_grd_print(s, grd_f, io_file, n)
        use ommp_interface
        implicit none
        
        type(ommp_system), intent(inout), target :: s
        !! System data structure
        procedure(grad_term), pointer :: grd_f 
        !! The analytical gradients function
        integer :: io_file
        !! File handler for I/O
        character(len=*) :: n
        !! Name of the component for I/O

        real(ommp_real), allocatable :: grad_ana(:,:)
        integer(ommp_integer) :: nat, i

        nat = s%top%mm_atoms

        allocate(grad_ana(3,nat))

        call grd_f(s, grad_ana)
        write(io_file, "('Grad ', A, ' ')") n
        do i=1, nat
            write(io_file, "(F20.8, ' ', F20.8, ' ', F20.8)") grad_ana(:,i) * OMMP_AU2KCALMOL * OMMP_ANG2AU
        end do
        write(io_file, *)
    end subroutine
end module

program test_geomgrad_xyz
    use iso_c_binding, only: c_char
    use iso_fortran_env, only: error_unit
    use ommp_interface
    use anagg_print

    implicit none

    integer :: argc, i, rc=0, fp=102
    character(kind=c_char, len=120), dimension(3) :: argv
    real(ommp_real), allocatable :: ef(:,:), ef_pol(:,:), &
                                    grad_num(:,:), grad_ana(:,:), delta(:,:)
    type(ommp_system), pointer :: my_system
    procedure(grad_term), pointer :: grd_f

    argc = command_argument_count()
    if(argc /= 4 .and. argc /= 3 ) then
        write(*, *) "Syntax expected"
        write(*, *) "    $ test_init.exe <XYZ FILE> <PRM FILE> <OUTPUT FILE>&
                    & [<ELECTRIC FIELD FILE>]"
        stop 1
    end if

    do i=1, argc
        call get_command_argument(i, argv(i))
    end do

    call ommp_init_xyz(my_system, argv(1), argv(2))

    allocate(ef(3, my_system%top%mm_atoms))
    allocate(ef_pol(3, my_system%eel%pol_atoms))
    
    if(argc == 4) then
        write(*, *) "Currently not supported!"
        write(error_unit, *) 1
        stop 1
    else
        ef = 0.0
    end if

    do i=1, my_system%eel%pol_atoms
        ef_pol(:,i) = ef(:, my_system%eel%polar_mm(i))
    end do
    
    deallocate(ef)
    
    open(fp, file=argv(3))
    grd_f => ommp_full_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "ETOT")
    
    grd_f => ommp_vdw_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EV")
    
    grd_f => ommp_fixedelec_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EM")
    
    grd_f => ommp_polelec_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EP")
    
    grd_f => ommp_bond_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EB")
    
    grd_f => ommp_torsion_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "ET")
    
    grd_f => ommp_angle_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EA")
    
    grd_f => ommp_urey_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EUB")
    
    grd_f => ommp_imptorsion_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EIT")
    
    grd_f => ommp_strbnd_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EBA")
    
    grd_f => ommp_angtor_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EAT")
    
    grd_f => ommp_opb_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EOPB")
    
    grd_f => ommp_strtor_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EBT")
    
    grd_f => ommp_tortor_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "ETT")
    
    grd_f => ommp_pitors_geomgrad
    call ana_grd_print(my_system, grd_f, fp, "EPT")
    close(fp)
    if(rc > 0) then
        write(error_unit, *) 1
        stop 1
    else
        write(error_unit, *) 0
        stop 0
    end if
end program
